plugins/modules/route53.py

```diff
@@ -593,113 +509,106 @@ def main():
+            ns = route53.get_hosted_zone(Id=zone_id)['DelegationSet']['NameServers']
 
-        module.exit_json(changed=False, set=record, nameservers=ns)
+        module.exit_json(changed=False, set=aws_record, nameservers=ns)
 
-    if command_in == 'delete' and not found_record:
+    if command_in == 'delete' and not aws_record:
         module.exit_json(changed=False)
 
-    changes = ResourceRecordSets(conn, zone_id)
-
     if command_in == 'create' or command_in == 'delete':
-        if command_in == 'create' and found_record:
+        if command_in == 'create' and aws_record:
             if not module.params['overwrite']:
                 module.fail_json(msg="Record already exists with different value. Set 'overwrite' to replace it")
             command = 'UPSERT'
         else:
             command = command_in.upper()
-        # Restore original order of records
-        wanted_rset.resource_records = unsorted_records
-        changes.add_change_record(command, wanted_rset)
 
     if not module.check_mode:
         try:
-            invoke_with_throttling_retries(commit, changes, retry_interval_in, wait_in, wait_timeout_in)
-        except boto.route53.exception.DNSServerError as e:
-            txt = e.body.split("<Message>")[1]
-            txt = txt.split("</Message>")[0]
-            if "but it already exists" in txt:
+            change_resource_record_sets = route53.change_resource_record_sets(
+                HostedZoneId=zone_id,
+                ChangeBatch=dict(
+                    Changes=[
+                        dict(
+                            Action=command,
+                            ResourceRecordSet=resource_record_set
+                        )
+                    ]
+                )
+            )
+
+            if wait_in:
+                waiter = route53.get_waiter('resource_record_sets_changed')
+                waiter.wait(
+                    Id=change_resource_record_sets['ChangeInfo']['Id'],
+                    WaiterConfig=dict(
+                        Delay=WAIT_RETRY,
+                        MaxAttemps=wait_timeout_in // WAIT_RETRY,
+                    )
+                )
+        except route53.exceptions.InvalidChangeBatch as e:
+            if "but it already exists" in e.response['Error']['Message']:
                 module.exit_json(changed=False)
             else:
-                module.fail_json(msg=txt)
-        except TimeoutError:
-            module.fail_json(msg='Timeout waiting for changes to replicate')
+                module.fail_json(msg=e)
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg=e)
+        except botocore.exceptions.WaiterError as e:
+            if "Max attempts exceeded" in e.kwargs['reason']:
+                module.fail_json(msg='Timeout waiting for changes to replicate', last_response=e.last_response)
+
+            module.fail_json(msg='Error on waiting for reply', last_response=e.last_response)
```

Comment:
We have some helpers which make handling the botocore errors much easier.

```
from ansible_collections.amazon.aws.plugins.module_utils.ec2 import is_boto3_error_message
```


This then lets you do:

```
except is_boto3_error_message('but it already exists'):
    module.exit_json(changed=False)
except botocore.exceptions.WaiterError as e:
    ...
except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
    module.fail_json_aws(e, msg='Failed to update records')
```

There's also `is_boto3_error_code`.

Boto3 / BotoCore exceptions are somewhat inconsistent in their implementation.  **But** all of the exceptions thrown as a result an API request will include the error Code and Message.  Unlike the boto3 documentation, the API documentation includes the codes that will be thrown.  is_boto3_error_code will let you check the codes (while not needing to know the exact exception - which isn't always consistently named or located).

See also https://docs.ansible.com/ansible/devel/dev_guide/platforms/aws_guidelines.html#using-is-boto3-error-code

We generally use is_boto3_error_code / is_boto3_error_message rather than trying to catch the specific exceptions as it's more consistent in the implementation.